### PR TITLE
HAL_ChibiOS: Add QioTekZealotF427 support

### PR DESCRIPTION
**Pull request by QioTek**
HAL_ChibiOS: Add QioTekZealotF427 support

the hwdef to support the QioTekZealotF427
It has 3 IMU sensors(ICM20698,ICM20602,BMI055), a Compass(Ist8310), 2Baro sensors(DPS310,MS5611), an integrated analog OSD, 14motor outputs, 4Relay output, 5uarts, 2Power Modules input, an i2c, a Can. 

I've build and tested for this specific target on master :
Bootloader
Plane
Copter

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef-bl.dat [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# hw definition file for processing by chibios_hwdef.py
+# for for QioTekZealotF427 hardware bootloader from Qio-tek.com 
+
+MCU STM32F4xx STM32F427xx
+
+APJ_BOARD_ID 1021
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef-bl.dat:6]: You need to add a new board type to https://github.com/ArduPilot/ardupilot/blob/master/Tools/AP_Bootloader/board_types.txt

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat [hunk 2]
```diff
@@ -0,0 +1,229 @@
+# hw definition file for processing by chibios_hwdef.py
+# for QioTekZealotF427 hardware from Qio-tek.com 
+
+# MCU class and specific type
+MCU STM32F4xx STM32F427xx
+
+# board ID for firmware load
+APJ_BOARD_ID 1301
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:8]: Needs to be added to board_types.txt

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+# 6) SERIAL5: extra UART or sbus output (usually RTOS debug console)
+
+SERIAL_ORDER OTG1 USART2 USART1 USART3 UART4 UART5
+
+# UART1 
+PB6 USART1_TX USART1
+PB7 USART1_RX USART1
+
+# USART2 for Mavlink wifi module set baudrate to 921600
+PD5 USART2_TX USART2 NODMA
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:51]: Disabling DMA on most of the UARTS doesn't seem great to me - is it necessary?

**Review by andyp1per — CHANGES_REQUESTED** [review]

**Comment by ShenMiaozhang** [comment]
Good board.

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat (hunk 2 above)
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:8]: OK, thanks for your prompt.

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat (hunk 3 above)
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:51]: It's necessary,STM32F427VIT6 used for the target.  Almost all the resources are used up.

**Comment by GundamF91C** [comment]
Looks good.Can't wait to try.When can I buy it?

**Comment by QioTek** [comment]
> Looks good.Can't wait to try.When can I buy it?

It will comeing soon, you can buy it on banggood.

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:8]: I think this should just be 1021 like the bootloader

Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:8]: sorry. my mistake, thanks.

**Review by andyp1per — APPROVED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat (hunk 2 above)
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat]: Sorry. It was my test id, I forgot to modify it.

**Comment by zhouzhiwen2000** [comment]
It's a very well-known and popular hobbyist flight controller in China. Can't wait for it to be mainlined.

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+define OSD_ENABLED 1
+define HAL_OSD_TYPE_DEFAULT 1
+ROMFS_WILDCARD libraries/AP_OSD/fonts/font*.bin
+
+# drdy pins
+PA15  DRDY_ICM20689 INPUT
+PA10  DRDY_ICM20602 INPUT
+# PD11 DRDY_ICM20649 INPUT
+
+#define AP_FEATURE_BRD_PWM_COUNT_PARAM 1
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:228]: AP_FEATURE_BRD_PWM_COUNT_PARAM line can be removed

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
we also would like this to follow the commit message conventions, please use:
  HAL_ChibiOS: xxxxx
for things in libraries/AP_HAL_ChibiOS
Please also add the board to Tools/scripts/build_binaries.py so the board will be automatically built by our build server

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+# UART1 
+PB6 USART1_TX USART1
+PB7 USART1_RX USART1
+
+# USART2 for Mavlink wifi module set baudrate to 921600
+PD5 USART2_TX USART2 NODMA
+PD6 USART2_RX USART2
+
+define HAL_SERIAL2_PROTOCOL SerialProtocol_MAVLink
```
[the remark this answers is not in the document]
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:54]: Sorry, the note copy from other target code, Just I forgot to change it. Usart2 commended use ESP8266 , and we will present it to this Flightcontroler for free.

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+# enable FAT filesystem support
+define HAL_OS_FATFS_IO 1
+# now some defines for logging and terrain data files
+define HAL_BOARD_LOG_DIRECTORY "/APM/LOGS"
+define HAL_BOARD_TERRAIN_DIRECTORY "/APM/TERRAIN"
+
+# define the order that I2C buses 
+I2C_ORDER I2C2 I2C1
+# this defines the default maximum clock on I2C devices.
+define HAL_I2C_MAX_CLOCK 100000
```
[the remark this answers is not in the document]
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:134]: the base code quote from F4BY. OK, I'll delete it.

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+PD15 TIM4_CH4 TIM4 PWM(11) GPIO(60)
+PD14 TIM4_CH3 TIM4 PWM(12) GPIO(61)
+PC9  TIM3_CH4 TIM3 PWM(13) GPIO(62) NODMA
+PC8  TIM3_CH3 TIM3 PWM(14) GPIO(63) NODMA
+
+PC14 EXTERN_GPIO1 GPIO(64)
+PC13 EXTERN_GPIO2 GPIO(65)
+PE4  EXTERN_GPIO3 GPIO(66)
+PE6  EXTERN_GPIO4 GPIO(67)
+define RELAY1_PIN_DEFAULT 64
```
[the remark this answers is not in the document]
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:167]: Yes, I ignored use for RPM or wheel encoders, I will modify it.

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat (hunk 4 above)
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:228]: OK, thanks.

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+# enable RAMTROM parameter storage
+define HAL_WITH_RAMTRON 1
+# enable FAT filesystem support
+define HAL_OS_FATFS_IO 1
+# now some defines for logging and terrain data files
+define HAL_BOARD_LOG_DIRECTORY "/APM/LOGS"
+define HAL_BOARD_TERRAIN_DIRECTORY "/APM/TERRAIN"
+
+# define the order that I2C buses 
+I2C_ORDER I2C2 I2C1
```
[the remark this answers is not in the document]
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:132]: It has been modified.

**Review by QioTek — COMMENTED** [review]
It has been modified.

**Comment by bys1123** [comment]
Wow! Nice work! Looking forward this board!

**Comment by tridge** [comment]
@QioTek I've fixed up your commit messages, squashing them and fixing messages. I've also added the bootloader to the build

**Review by tridge — APPROVED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef-bl.dat [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+OSCILLATOR_HZ 24000000
+
+STM32_ST_USE_TIMER 5
+
+# flash size
+FLASH_SIZE_KB 2048
+FLASH_RESERVE_START_KB 0
+
+# the location where the bootloader will put the firmware
+FLASH_BOOTLOADER_LOAD_KB 128
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef-bl.dat:17]: why do you reserve so much space for the bootloader? The minimum would be 16k. Using 128k wastes quite a lot of flash space. You have plenty now, but it does mean as we get close to filling 2MByte you won't be able to have as many features

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+# for QioTekZealotF427 hardware from Qio-tek.com 
+
+# MCU class and specific type
+MCU STM32F4xx STM32F427xx
+
+# board ID for firmware load
+APJ_BOARD_ID 1021
+
+FLASH_SIZE_KB 2048
+FLASH_RESERVE_START_KB 128
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:11]: if you change flash start you'll need to change it here too

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat (hunk 10 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat:11]: Jason has told me he wants to keep 128k

**Review by tridge — APPROVED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef-bl.dat (hunk 9 above)
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef-bl.dat:17]: Flash after the current firmware compiled a lot free. so I kept the boot loader 128K. but I will I will modify it.

File: libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat (hunk 10 above)
Comment (QioTek) [reply libraries/AP_HAL_ChibiOS/hwdef/QioTekZealotF427/hwdef.dat]: I'm modified to16K.

**Comment by QioTek** [comment]
@Hwurzburg I posted Basic introduction.

**Pull request merged**